helix-core/src/test/java/org/apache/helix/manager/zk/TestRawZkClient.java

```diff
@@ -461,4 +467,247 @@ void testPendingRequestGauge() throws Exception {
+    Stat stat = new Stat();
+    String nodeData = _zkClient.readData(path, stat, true);
+
+    Assert.assertNotNull(nodeData, "Failed to create ephemeral node: " + path);
+    Assert.assertEquals(nodeData, data, "Data is not correct.");
+    Assert.assertTrue(stat.getEphemeralOwner() != 0L,
+        "Ephemeral owner should NOT be zero because the node is an ephemeral node.");
+    Assert.assertEquals(ZKUtil.toHexSessionId(stat.getEphemeralOwner()), originalSessionId,
+        "Ephemeral node is created by an unexpected session");
+
+    // Delete the node to clean up, otherwise, the ephemeral node would be existed
+    // until the session of its owner expires.
+    _zkClient.delete(path);
+  }
+
+  /*
+   * This test checks that ephemeral creation fails because the expected zk session does not match
+   * the actual zk session.
+   * How this test does is:
+   * 1. Creates a zk client and gets its original session id
+   * 2. Expires the original session, and a new session will be created
+   * 3. Tries to create ephemeral node with the original session
+   * 4. ZkSessionMismatchedException is expected for the creation and ephemeral node is not created
+   */
+  @Test
+  public void testCreateEphemeralWithMismatchedSession() throws Exception {
+    final String className = TestHelper.getTestClassName();
+    final String methodName = TestHelper.getTestMethodName();
+    final String clusterName = className + "_" + methodName;
+
+    TestHelper.setupCluster(clusterName, ZK_ADDR,
+        12918, // participant port
+        "localhost", // participant name prefix
+        "TestDB", // resource name prefix
+        1, // resources
+        10, // partitions per resource
+        5, // number of nodes
+        3, // replicas
+        "MasterSlave",
+        true); // do rebalance
+
+    final long originalSessionId = _zkClient.getSessionId();
+    final String originalHexSessionId = _zkClient.getHexSessionId();
+    final String path = "/" + methodName;
+
+    // Verify the node is not existed.
+    Assert.assertFalse(_zkClient.exists(path));
+
+    // Expire the original session.
+    ZkTestHelper.expireSession(_zkClient);
+
+    // Wait until the ZkClient has got a new session.
+    Assert.assertTrue(TestHelper.verify(() -> {
+      try {
+        // New session id should be greater than expired session id.
+        return _zkClient.getSessionId() > originalSessionId;
+      } catch (HelixException ex) {
+        return false;
+      }
+    }, 1000L));
+
+    try {
+      // Try to create ephemeral node with the original session.
+      // This creation should NOT be successful because the original session is already expired.
+      _zkClient.createEphemeral(path, "Hello Helix", originalHexSessionId);
+      Assert.fail("Ephemeral node should not be created by the expired session.");
+    } catch (ZkSessionMismatchedException expected) {
+      // Expected because there is a session mismatch.
+    }
```

Comment:
The purpose of this test is to test a mismatched session won't create an ephemeral node. With the check logic, I think ZkSessionMismatchedException is expected. So I don't think we need to fail this test if any other exception is thrown.

Reply:
` try { }
catch (ZkSessionMismatchedException expected) {
} 
catch (Exception eNotExpected) {
Assert.fail("unexpected exception" + e);
}`
If other exception happens, there is something wrong. You want to catch it. Otherwise, it is deemed as right, which is not the case.

Reply:
It is fine. Added. Thanks.